### ISSUE TITLE
New topic: Web API +  authentication and authorization

### ISSUE THREAD

**Issue by Rick-Anderson** [aspnet/AspNetCore.Docs#4556]
We've had several [customer requests for this](https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-web-api?hubRefSrc=email&utm_source=lfemail&utm_medium=email&utm_campaign=lfnotification#lf-content=177016560:743571877&reply_editor=true).
Related: Topic: how to set up authentication for ASP.NET Core Web API with ADFS #3845

[From Web API tutorial](https://docs.microsoft.com/en-us/aspnet/core/tutorials/first-web-api)
>Everything works great and I must say amazing job in step by step explanation. Hope this sample with authentication and authorization will be up soon. Btw, it would be great if on one example we make full circle, create API (this tutorial), add authentication second, then deploy to Azure or other platform and then consume from other application.E

@danroth27  can you assign someone to write a sample and outline the topic?

cc @scottaddie 
WebAPI

**Comment by isaac2004** [comment]
is this only available with Identity Server 4? or is there something built-in now?

**Comment by Rick-Anderson** [comment]
@danroth27 can you assign someone to write a sample and outline the topic?

**Comment by cwoodruff** [comment]
I can do it and have talked to @scottaddie about it

**Comment by scottaddie** [comment]
@danroth27 Do you have an outline in mind, or should @cwoodruff take a shot at that?

**Comment by danroth27** [comment]
We certainly write topics about securing Web APIs using Azure AD, Azure AD B2C and ADFS. We should look first to see if there are existing Azure topics that we can just reference. For ADFS I suspect we'll need a new topic.

We don't currently have a story for issuing tokens from an ASP.NET Core app. For that you need to use something like IdentityServer4. Folks wanting to use IdentityServer4 with ASP.NET Core should refer to the [IdentityServer4 docs](http://docs.identityserver.io/en/release/), which we already link to.

**Comment by Rick-Anderson** [comment]
See [forum post](https://forums.asp.net/post/6156135.aspx)

@scottaddie  can you follow up with @cwoodruff  on this?

**Comment by isaac2004** [comment]
@cwoodruff @scottaddie did this doc ever get written, I can help out if needed.

**Comment by scottaddie** [comment]
@cwoodruff Please let us know if you still have interest and time to write this doc.

@isaac2004 This hasn't been written yet.

**Comment by Rick-Anderson** [comment]
@cwoodruff Please let us know if you still have interest and time to write this doc.

**Comment by Rick-Anderson** [comment]
@CamSoper  would you be able to work on any of securing Web APIs using Azure AD, Azure AD B2C and ADFS?

**Comment by CamSoper** [comment]
@Rick-Anderson Yes, I have recent experience there!  I can help with that.

**Comment by Rick-Anderson** [comment]
@CamSoper  can you give us an ETA for starting this work?

**Comment by CamSoper** [comment]
@Rick-Anderson I'm loaded up for the rest of this sprint, but if you want to assign it for 140, I can get to it then.

**Comment by Rick-Anderson** [comment]
@CamSoper  scheduled for 141

**Comment by CamSoper** [comment]
Thanks, @Rick-Anderson .  Sorry it's been taking me so long.  I've had a lot of irons in a lot of fires these past couple sprints.

**Comment by Rick-Anderson** [comment]
@CamSoper 	moved to 142

**Comment by CamSoper** [comment]
@isaac2004 This has been assigned to me in the backlog for a while, it's time we got it done. You still up for helping out?  It'd be fun to collaborate!  HMU on Teams.

**Comment by CamSoper** [comment]
@danroth27 - Hey Dan, @isaac2004 and I were just taking a look at this and we have questions.  Can you clarify the ask here?

Are we looking specifically for a topic wiring up an ASP.NET Core Web API and ADFS specifically?  [Vittorio has that scenario on his blog](http://www.cloudidentity.com/blog/2014/02/12/use-the-on-premises-organizational-authentication-option-adfs-with-asp-net-in-visual-studio-2013/), albeit it's a bit older and doesn't work for ASP.NET Core as written.

Basically I'm asking if the object of the doc is to wire up a Web API with a generic OAuth authentication provider (like Azure AD, ADFS, etc.) **or** if the doc is to wire up a Web API with ADFS and *only* ADFS.

And how much of a client do we want?  SPA?

**Comment by isaac2004** [comment]
BTW, the extension that seems to possibly do this is

```csharp
  services.AddAuthentication(AzureADDefaults.AuthenticationScheme)
            .AddWsFederation(options => new WsFederationOptions
            {
                Wtrealm = Configuration["ida:Wtrealm"],
                MetadataAddress = Configuration["ida:ADFSMetadata"]
            });
````

Would need to test by spinning up a DC however

**Comment by danroth27** [comment]
@CamSoper Please work with @blowdart on this. There is also relevant work happening in ASP.NET Core 2.2 around Web API authentication that you'll probably want to take into account.

**Comment by CamSoper** [comment]
FYI everybody (esp. @danroth27 and @rick-anderson) - @blowdart and I decided that this topic should wait until 2.2 is near release because 2.2 will have some critical new functionality in this regard.  Barry says we should focus on AAD and AAD B2C scenarios first -- Which happens to be the other issues assigned to me, so that's what I'll be doing in the short term.

**Comment by dmitry-pavlov** [comment]
@Rick-Anderson are there any updates? 

I need to authenticate and authorize the user inside ASP.NET Core 2.2 Web API which is deployed to GKE (Google Kubernetes Engine) as docker container and is being called by UI app as well as by other other APIs (microservices). In request header a bearer token passed to my API. So I need to identify the user the request sent on behalf of, and figure out the user permissions. I also will need to pass the token to other APIs when I call them via HttpClient. The users and their permissions come from ADFS and login process is SSO. 

Any pointers to any helpful materials are nuch appreciated. Thanks!

**Comment by danacon01** [comment]
This topic seems to be focusing on guidance using heavyweight directories such as AD and Azure AD, but I would like to put my vote in for including guidance for securing APIs with authentication/authorisation that don't have these services supporting them. Maybe using a custom user repository like the SQL one that comes by default with Identity (that may or may not be appropriate, but then that is why I am asking for guidance...).

My personal scenario is for securing web api's that are internet facing that are being consumed by client systems not users/web apps (I guess this would be a B2B integration scenario).

**Comment by Rick-Anderson** [comment]
Per `@HaoK`
There's a bit of conflation here, and there are two parts of identity, users/roles and that stuff is applicable everywhere, identity by default uses cookies to store that info, so its that piece that is not applicable to WebAPI, so they would have to do the work to use identity to login/etc, and then typically generate a bearer token (which is why IdentityServer/OpenIdict is used there).  

We are also integrating identity server with identity:
https://github.com/aspnet/AspNetCore/issues/5833

**Comment by danacon01** [comment]
@Rick-Anderson so is there guidance anywhere for the first part you mentioned i.e. use identity to login to a web api?

i think it should be reasonably straightforward to wire in the non-ui bits of identity to a web api project to get the sql backend in place, but i am wondering if there is inbuilt framework stuff or guidance for the logging in bit or whether i need write code for example for grabbing an http authorisation header (e.g. client using basic auth over https) and running some sort of validateuser method on the extracted/decoded creds (which feels a bit old skool). i have seen a few different articles on the web, just not sure what is the best way for my scenario.

**Comment by wewald** [comment]
I've watched Dan Roth and Chris Woodruff's you tube video's to get us going with Core's Web API's but we have to authenticate via .net Identity (which our current .net 4.6.1 Azure website uses).   Rick Anderson seemed to say to wait until the beginning of 2019.  Can you point me to how to do this?

BTW - I also commented on this issue at:
https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity?view=aspnetcore-2.2&tabs=visual-studio#feedback

**Comment by cwoodruff** [comment]
@wewald thanks for watching the videos. I am planning on getting some blog posts and videos developed covering AuthN and AuthZ for ASP.NET Core Web API. If you can use a third-party solution, I would recommend looking at Auth0 https://auth0.com/.

**Comment by mkArtakMSFT** [comment]
@javiercn are there docs covering this already?

**Comment by Rick-Anderson** [comment]
@cwoodruff  or @CamSoper  - any chance one of you could work on this?

**Comment by javiercn** [comment]
@mkArtakMSFT This is for on premise active directory (not Azure Active Directory) which is not something we have out of the box support. I've googled it and found the following docs:
https://docs.microsoft.com/en-us/windows-server/identity/ad-fs/development/single-page-application-with-ad-fs

I recommend @danroth27 or @Rick-Anderson follow up with the Identity group doc folks to either update their docs for asp.net core or put together a separate sample. Ultimately it would mean to change the code to use ASP.NET Core.

@mkArtakMSFT I don't think we have to take any action here. So we should either close it or move it to the appropriate docs repo. There is a github repo for something like it https://github.com/microsoft/adfs-sample-msal-dotnet-webapp-to-webapi and I'm not sure if there is a more appropriate repo or who to reach out. 

@danroth27 @Rick-Anderson Do any of you know the right folks to reach out to?

**Comment by javiercn** [comment]
@mkArtakMSFT I consider my part in this handled, as this scenario is not something our team directly owns.

**Comment by mkArtakMSFT** [comment]
@blowdart this seems to be a highly requested topic to document. Do you want to do that? Or maybe @anurse can suggest somebody to to handle this?

**Comment by blowdart** [comment]
No. If it's ADFS then it should probably fall to the identity / ADAL team

**Comment by Rick-Anderson** [comment]
I created https://github.com/MicrosoftDocs/windowsserverdocs/issues/3312 and reached out to the team.

**Comment by wewald** [comment]
In seeing the possibility of getting some traction on this item, I wanted to clarify that the need we  see is to extend website authentication currently done via Identity (2019 link below) to authenticate RestAPIs users.  My understanding is that OAuth is needed to do this, which would not seem to have any need to  involve Active Directory Federation Services (which to my understanding is to use Azure's Active Directory database as opposed to the SQL tables that are built as part of Identity).

https://docs.microsoft.com/en-us/aspnet/core/security/authentication/identity

**Comment by Rick-Anderson** [comment]
> the need we see is to extend website authentication currently done via Identity (2019 link below) to authenticate RestAPIs users.

That's what the issue was opened on. I think the issue got hijacked by ADFS.  @blowdart  @javiercn  please respond.

**Comment by blowdart** [comment]
Well now you're changing what the original request was for. The original request was ADFS. Using identity is something different and involves identity server, again, a third party product and not something we would document.

**Comment by wewald** [comment]
"Identity" is not a 3rd party product.  

"The ASP.NET membership system was introduced with ASP.NET 2.0 back in 2005, and since then there have been many changes in the ways web applications typically handle authentication and authorization. ASP.NET Identity is a fresh look at what the membership system should be when you are building modern applications for the web, phone, or tablet." https://docs.microsoft.com/en-us/aspnet/identity/overview/getting-started/introduction-to-aspnet-identity

**Comment by Rick-Anderson** [comment]
@wewald  see https://identityserver.io/ which is **not** a Microsoft product.

**Comment by danacon01** [comment]
chiming in here again as i agree with wewald that the request for guidance is nothing to do with adfs. there is guidance in the docs for how to do this with aspnet mvc but there is no comparable guidance for web api (which is why the request materialised in the first place i think).
@Rick-Anderson you pointed us to identityserver, are you saying then that there is no solution in aspnet core web api and identity for authenticating web api's and that our only recourse is a third party product?

**Comment by wewald** [comment]
I am very surprised to learn about this product ambiguity as we got started with Membership in  2006 and are of the perspective that we need to move to Identity to align what we want to be an  Azure based strategy and we now need to take orders via RestAPIs to our offerings.  

The link below has a list of .net Framework support from 2.0-4.8.   We need guidance from Microsoft on how we extend the path Microsoft guided us to, to that that allows us to use RestAPIs.

https://docs.microsoft.com/en-us/dotnet/api/system.web.security.membership

**Comment by Rick-Anderson** [comment]
Fixed in #15288
